### PR TITLE
#221 Initial design for directory functionality

### PR DESCRIPTION
**Pull request by pinaf**
For #221
- added  class `TkDirCreate` + unit test
- `TkDir` now takes instance of `Base` in constructor
- added interfaces `Dir` and `Dirs`
- added puzzle for other directory operations
- added puzzle for creation of a `RqDir` class for bridge between request and `Dir`

### CODE REVIEW

**Comment by pinaf** [comment]
@karato travis failure is unrelated to issue. see #299

**Comment by karato** [comment]
@pinaf Thanks for your pull request, let me find someone who can review it

**Comment by karato** [comment]
@krzyk review it

File: src/main/java/com/nerodesk/takes/doc/TkDirCreate.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * Ctor.
+     * @param bse Base.
+     */
+    TkDirCreate(final Base bse) {
+        this.base = bse;
+    }
+
+    // @todo #221:30min This is a stub implementation. We should create a class
+    //  RqDir, analogous to RqDoc, to handle the actual directory creation.
```
Comment (krzyk) [inline src/main/java/com/nerodesk/takes/doc/TkDirCreate.java]: @pinaf here you should mention `Dir` and `Dirs` interfaces, right now they are created but are not used

**Comment by krzyk** [comment]
@pinaf I have one comment regarding the todo

**Comment by pinaf** [comment]
@krzyk ok changed it. please take another look.

**Comment by krzyk** [comment]
@karato we are waiting for #299

**Comment by karato** [comment]
> @karato we are waiting for #299 

@krzyk yes, of course, we'll wait for #299

**Comment by krzyk** [comment]
@pinaf qulice fails with:

```
[ERROR] /src/main/java/com/nerodesk/takes/doc/TkDirCreate.java[50-50]: Perhaps 'base' could be replaced by a local variable. (SingularField)

[ERROR] /src/main/java/com/nerodesk/takes/doc/TkDirCreate.java[50-50]: Avoid unused private fields such as 'base'. (UnusedPrivateField)
```

**Comment by karato** [comment]
@pinaf we waited for #299, it is closed already

**Comment by k0r** [comment]
@yegor256 we might have to postpone this feature. directory implementation is going cause a few code challenges: directory deletion, moving files to another directory, and reorganizing files if the directory is deleted, etc. could be several tickets to scale this feature.

**Comment by k0r** [comment]
@yegor256 what do you think?

**Comment by yegor256** [comment]
@karato indeed, let's postpone this

**Comment by karato** [comment]
> @karato indeed, let's postpone this

@yegor256 OK, I labeled it as "postponed" tag

**Comment by karato** [comment]
> @karato indeed, let's postpone this

@yegor256 right, I'll try to find someone else for this task

**Comment by karato** [comment]
@longtimeago review it

**Comment by pinaf** [comment]
@k0r @yegor256 not sure this was intentional or not, but ever since I have been added to the nerodesk team I have permission to merge the pull request...

**Comment by k0r** [comment]
@pinaf `nerodesk` team has `write` permission.

**Comment by pinaf** [comment]
@k0r so it was intentional?

**Comment by k0r** [comment]
@pinaf #271

**Comment by yegor256** [comment]
@k0r please fix it. the team must have "read-only" permissions

**Comment by k0r** [comment]
@yegor256 fixed.

**Comment by yegor256** [comment]
@k0r you can also remove me and @karato from nerodesk-read team

**Comment by k0r** [comment]
@karato please, be careful with the labels.

**Comment by pinaf** [comment]
@yegor256 shall we proceed with CR ?

**Comment by longtimeago** [comment]
@pinaf good question .. I'm also waiting for the answer

**Comment by krzyk** [comment]
@yegor256 and why was it reassigned? I think it should have been unassigned completely until @k0r decides we should go with directories

**Comment by yegor256** [comment]
@longtimeago I think we should proceed, review and merge. It's a pull request, we shouldn't hold it open for long time. (no matter what the status of the project is)

**Comment by k0r** [comment]
@yegor256 what about [this](https://github.com/Nerodesk/nerodesk/pull/300#issuecomment-107742519)?

**Comment by yegor256** [comment]
@k0r my bad, I thought it's an issue, not a pull request. It's not a good practice to keep pull requests in "on hold" status for longer. It will "expire" and we'll seriously damage the motivation of its author. It's always better to merge, no matter how "bad" is the code (well, it has to pass rultor and code review, of course).

**Comment by k0r** [comment]
@yegor256 postponing the feature isn't intended to demotivate the programmer, it involves other challenges we've previously discussed [here](https://github.com/Nerodesk/nerodesk/pull/300#issuecomment-107735787).

**Comment by yegor256** [comment]
@k0r yes, postponing a feature is perfectly fine, but it's a pull request, not a feature. there is already some code written.

**Comment by k0r** [comment]
@yegor256 thanks. proceed, if needed.

**Comment by krzyk** [comment]
@yegor256 in such case can we reassign it to back to me,  as I already reviewed it and,  the only thing that is left is saying "merge"  :-)

**Comment by longtimeago** [comment]
@krzyk I don't mind :)

**Comment by pinaf** [comment]
@karato please assign @krzyk here

**Comment by pinaf** [comment]
@yegor256 could you assign @krzyk please?

**Comment by yegor256** [comment]
@karato assign @krzyk to this pls

**Comment by karato** [comment]
> @karato assign @krzyk to this pls

@yegor256 OK. @krzyk the ticket is yours, please proceed

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by yegor256** [comment]
@rultor merge

**Comment by karato** [comment]
@krzyk Thanks a lot, I just topped your account for **21 mins**, transaction ID `58657984`, total time was 1 hour and 5 mins... extra minutes for review comments (c=6)... +21 added to your rating, at the moment it is: [+4684](http://www.netbout.com/b/35035?open=rating)

**Comment by karato** [comment]
@rultor deploy now pls

**Pull request merged**